packages/server/src/controllers/user.ts

```diff
@@ -1,29 +1,85 @@
-import { Request, Response } from 'express';
-import * as qs from 'qs';
+import { Request, Response } from "express";
+import * as uuid from "uuid/v4";
+import { URLSearchParams } from "url";
+import * as qs from "qs";
+import fetch from "node-fetch";
+import { isProduction } from "../utils";
+
+const SPOTIFY_STATE_KEY = "SPOTIFY_STATE_KEY";
+
+const baseUrl = isProduction() ? process.env.APP_URL : "http://localhost:4000";
+
+const redirectUri = `${baseUrl}/auth/callback`;
 
 export const login = (req: Request, res: Response) => {
-  const state = 'what is state';
-  const scope = 'user-read-private user-read-email';
+  const state = uuid();
+  const scope = "user-read-private user-read-email";
 
-  res.cookie('spotifyState', state);
+  res.cookie(SPOTIFY_STATE_KEY, state);
 
   const query = qs.stringify({
-    response_type: 'code',
+    response_type: "code",
     client_id: process.env.SPOTIFY_CLIENT_ID,
-    redirect_uri: 'http://localhost:4000/auth/callback',
+    redirect_uri: redirectUri,
     scope,
     state
   });
 
   res.redirect(`https://accounts.spotify.com/authorize?${query}`);
 };
 
-export const loginCallback = (req: Request, res: Response) => {
-  const code = req.query.code || null;
+export const loginCallback = async (req: Request, res: Response) => {
+  const { code, state } = req.query;
+  const storedState = req.cookies?.SPOTIFY_STATE_KEY ?? null;
 
-  if (code) {
-    return res.status(200).send('You have a code 🥳');
+  if (!code || !state || state !== storedState) {
+    return res.status(404).send("😞");
```

Comment:
Changed from `403` to `404` for better security

Reply:
Good to know!